In [1]:
#IMPORTANDO LIBRERIAS NECESARIAS
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)

In [ ]:
# Traigo la información del csv

df_datos_abc = pd.read_csv("hr_raw_data.csv")
df_datos_abc

Unnamed: 0 age attrition businesstravel    dailyrate  \
0              0  51        No            NaN  2015.722222   
1              1  52        No            NaN  2063.388889   
2              2  42        No  travel_rarely  1984.253968   
3              3  47        No  travel_rarely  1771.404762   
4              4  46        No            NaN  1582.771346   
...          ...  ..       ...            ...          ...   
1673        1673  43        No            NaN   488.944444   
1674        1674  47        No            NaN  1973.984127   
1675        1675  29        No  travel_rarely   290.035510   
1676        1676  47        No  travel_rarely  1032.487286   
1677        1677  32        No            NaN   556.256661   

                    department  distancefromhome  education    educationfield  \
0                          NaN                 6          3               NaN   
1                          NaN                 1          4     Life Sciences   
2      Research & Development                  4          2  Technical Degree   
3                          NaN                 2          4           Medical   
4                          NaN                 3          3  Technical Degree   
...                        ...               ...        ...               ...   
1673                       NaN               -26          3           Medical   
1674                       NaN                26          4               NaN   
1675                       NaN                15          3               NaN   
1676                       NaN                 4          3     Life Sciences   
1677                       NaN                 2          2     Life Sciences   

      employeecount  employeenumber  environmentsatisfaction  gender  \
0                 1               1                        1       0   
1                 1               2                        3       0   
2                 1               3                        3       0   
3                 1               4                        1       1   
4                 1               5                        1       1   
...             ...             ...                      ...     ...   
1673              1             824                        2       1   
1674              1            1087                        4       1   
1675              1             528                        3       0   
1676              1              76                        3       1   
1677              1             401                        4       1   

      hourlyrate  jobinvolvement  joblevel                   jobrole  \
0            NaN               3         5        resEArch DIREcToR    
1            NaN               2         5                  ManAGeR    
2            NaN               3         5                  ManaGER    
3            NaN               3         4        ReseArCH DIrECtOr    
4            NaN               4         4          sAleS EXECUtIve    
...          ...             ...       ...                       ...   
1673         NaN               4         1       rESEaRcH SciEnTiST    
1674         NaN               3         5                  mANager    
1675         NaN               3         1       reSearch sCienTiSt    
1676         NaN               2         3   maNufACTURING DIREctOr    
1677   69.532083               3         2       resEArch scIENTisT    

      jobsatisfaction maritalstatus monthlyincome monthlyrate  \
0                   3           NaN     16280,83$   42330,17$   
1                   3           NaN           NaN   43331,17$   
2                   4       Married           NaN   41669,33$   
3                   3       Married     14307,50$   37199,50$   
4                   1      Divorced     12783,92$   33238,20$   
...               ...           ...           ...         ...   
1673                3        Single      3949,17$   10267,83$   
1674                3       Married   

In [3]:
df_datos_abc.columns

Index(['Unnamed: 0', 'age', 'attrition', 'businesstravel', 'dailyrate',
       'department', 'distancefromhome', 'education', 'educationfield',
       'employeecount', 'employeenumber', 'environmentsatisfaction', 'gender',
       'hourlyrate', 'jobinvolvement', 'joblevel', 'jobrole',
       'jobsatisfaction', 'maritalstatus', 'monthlyincome', 'monthlyrate',
       'numcompaniesworked', 'over18', 'overtime', 'percentsalaryhike',
       'performancerating', 'relationshipsatisfaction', 'standardhours',
       'stockoptionlevel', 'totalworkingyears', 'trainingtimeslastyear',
       'worklifebalance', 'yearsatcompany', 'yearsincurrentrole',
       'yearssincelastpromotion', 'yearswithcurrmanager',
       'sameasmonthlyincome', 'datebirth', 'salary', 'roledepartament',
       'numberchildren', 'remotework'],
      dtype='object')

In [5]:
df_carmen = df_datos_abc [['yearsatcompany', 'yearsincurrentrole',
       'yearssincelastpromotion', 'yearswithcurrmanager',
       'sameasmonthlyincome', 'datebirth', 'salary', 'roledepartament',
       'numberchildren', 'remotework']]

df_carmen

yearsatcompany yearsincurrentrole  yearssincelastpromotion  \
0                 20                NaN                       15   
1                 33                NaN                       11   
2                 22                NaN                       11   
3                 20                NaN                        5   
4                 19                NaN                        2   
...              ...                ...                      ...   
1673               3                NaN                        1   
1674               5                NaN                        1   
1675               6                NaN                        1   
1676              22                NaN                       14   
1677               8                NaN                        0   

      yearswithcurrmanager sameasmonthlyincome  datebirth      salary  \
0                       15           16280,83$       1972  195370,00$   
1                        9                 NaN       1971  199990,00$   
2                       15                 NaN       1981  192320,00$   
3                        6           14307,50$       1976  171690,00$   
4                        8           12783,92$       1977         NaN   
...                    ...                 ...        ...         ...   
1673                     2            3949,17$       1980         NaN   
1674                     0           15943,72$       1976  191324,62$   
1675                     5                 NaN       1994   28111,13$   
1676                    10            8339,32$       1976  100071,84$   
1677                     7                 NaN       1991   53914,11$   

                           roledepartament  numberchildren remotework  
0                                      NaN             NaN        Yes  
1                                      NaN             NaN          1  
2      ManaGER  -  Research & Development              NaN          1  
3                                      NaN             NaN      False  
4                                      NaN             NaN          0  
...                                    ...             ...        ...  
1673                                   NaN             NaN        Yes  
1674                                   NaN             NaN      False  
1675                                   NaN             NaN      False  
1676                                   NaN             NaN        Yes  
1677                                   NaN             NaN          0  

[1678 rows x 10 columns]

In [7]:
#- ¿Cuántas filas y columnas tenemos en el DataFrame?
print(f"tenemos {df_carmen.shape[0]} filas")
print(f"tenemos {df_carmen.shape[1]} columnas")

tenemos 1678 filas
tenemos 10 columnas


In [8]:
# 1. Muestra las primeras 5 filas del conjunto de datos
print("Primeras 5 filas:")
df_carmen.head()

Primeras 5 filas:


yearsatcompany yearsincurrentrole  yearssincelastpromotion  \
0              20                NaN                       15   
1              33                NaN                       11   
2              22                NaN                       11   
3              20                NaN                        5   
4              19                NaN                        2   

   yearswithcurrmanager sameasmonthlyincome  datebirth      salary  \
0                    15           16280,83$       1972  195370,00$   
1                     9                 NaN       1971  199990,00$   
2                    15                 NaN       1981  192320,00$   
3                     6           14307,50$       1976  171690,00$   
4                     8           12783,92$       1977         NaN   

                        roledepartament  numberchildren remotework  
0                                   NaN             NaN        Yes  
1                                   NaN             NaN          1  
2   ManaGER  -  Research & Development              NaN          1  
3                                   NaN             NaN      False  
4                                   NaN             NaN          0

In [9]:
# 2. Muestra las últimas 5 filas del conjunto de datos
print("Últimas 5 filas:")
df_carmen.tail()

Últimas 5 filas:


yearsatcompany yearsincurrentrole  yearssincelastpromotion  \
1673               3                NaN                        1   
1674               5                NaN                        1   
1675               6                NaN                        1   
1676              22                NaN                       14   
1677               8                NaN                        0   

      yearswithcurrmanager sameasmonthlyincome  datebirth      salary  \
1673                     2            3949,17$       1980         NaN   
1674                     0           15943,72$       1976  191324,62$   
1675                     5                 NaN       1994   28111,13$   
1676                    10            8339,32$       1976  100071,84$   
1677                     7                 NaN       1991   53914,11$   

     roledepartament  numberchildren remotework  
1673             NaN             NaN        Yes  
1674             NaN             NaN      False  
1675             NaN             NaN      False  
1676             NaN             NaN        Yes  
1677             NaN             NaN          0

In [10]:
# 3. Muestra una muestra aleatoria de 5 filas del conjunto de datos
print("Muestra aleatoria de 5 filas:")
df_carmen.sample(5)


Muestra aleatoria de 5 filas:


yearsatcompany yearsincurrentrole  yearssincelastpromotion  \
753                4                NaN                        1   
854                3                NaN                        1   
303                8                NaN                        7   
1442               0                NaN                        0   
912                2                NaN                        1   

      yearswithcurrmanager sameasmonthlyincome  datebirth      salary  \
753                      2            4492,84$       2000   53914,11$   
854                      2            2342,59$       1984   28111,13$   
303                      4            4856,67$       1997         NaN   
1442                     0            2342,59$       2005         NaN   
912                      2                 NaN       1977  105270,00$   

                                        roledepartament  numberchildren  \
753    laBorAtORy TECHnicIaN  -  Research & Developm...             NaN   
854                                                 NaN             NaN   
303                                                 NaN             NaN   
1442                                                NaN             NaN   
912    HeAlthcaRE rEPrEseNTaTIVE  -  Research & Deve...             NaN   

     remotework  
753           0  
854           0  
303           1  
1442       True  
912           1

In [11]:
# 4. Muestra el nombre de todas las columnas en el conjunto de datos
#Está todo en lowercase y sin espacios. No es necesario estandarizar
print("Nombre de todas las columnas:")
df_carmen.columns

Nombre de todas las columnas:


Index(['yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')

In [12]:
# 5. Muestra los valores únicos en la columna "employeenumber"
print("Valores únicos en la columna 'yearsatcompany':")
df_carmen['yearsatcompany'].unique()

Valores únicos en la columna 'yearsatcompany':


array([20, 33, 22, 19, 21, 18, 24, 31, 26, 16, 23, 15, 17, 32, 14, 13, 25,
       12, 11, 37, 40, 36, 27, 29, 10,  9, 30,  8,  7, 34,  6,  5,  4,  2,
        3,  1,  0])

In [13]:
# 6. Cuenta cuántas veces aparece cada valor único en la columna "employeenumber"
conteo_años = df_carmen['yearsatcompany'].value_counts()
print("Conteo de valores únicos en 'yearsatcompany':",conteo_años)

Conteo de valores únicos en 'yearsatcompany': yearsatcompany
5     214
1     181
3     149
2     145
10    139
4     116
7     115
8     113
9      98
6      84
0      47
11     37
20     30
13     27
15     22
14     19
22     18
18     16
12     15
16     14
21     14
19     12
17      9
24      7
25      5
33      5
26      4
36      4
32      3
27      3
31      3
40      3
29      2
23      2
34      1
37      1
30      1
Name: count, dtype: int64


In [14]:
# para verlos sin esos ... en el medio podriamos usar un ciclo for.

conteo_años = df_carmen['yearsatcompany'].value_counts()
for numero, count in conteo_años.items():
    print(f'{numero}: {count}')

5: 214
1: 181
3: 149
2: 145
10: 139
4: 116
7: 115
8: 113
9: 98
6: 84
0: 47
11: 37
20: 30
13: 27
15: 22
14: 19
22: 18
18: 16
12: 15
16: 14
21: 14
19: 12
17: 9
24: 7
25: 5
33: 5
26: 4
36: 4
32: 3
27: 3
31: 3
40: 3
29: 2
23: 2
34: 1
37: 1
30: 1


In [15]:
# 7. Verifica si hay valores nulos en el conjunto de datos

# usando info para ver mas datos
df_carmen.info()

# Verificar si hay valores no nulos en todo el DataFrame
df_carmen.isnull()

# Verificar si hay valores no nulos en una columna específica
#df['columna'].notnull()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1678 entries, 0 to 1677
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   yearsatcompany           1678 non-null   int64  
 1   yearsincurrentrole       35 non-null     object 
 2   yearssincelastpromotion  1678 non-null   int64  
 3   yearswithcurrmanager     1678 non-null   int64  
 4   sameasmonthlyincome      1189 non-null   object 
 5   datebirth                1678 non-null   int64  
 6   salary                   1393 non-null   object 
 7   roledepartament          312 non-null    object 
 8   numberchildren           0 non-null      float64
 9   remotework               1678 non-null   object 
dtypes: float64(1), int64(4), object(5)
memory usage: 131.2+ KB


yearsatcompany  yearsincurrentrole  yearssincelastpromotion  \
0              False                True                    False   
1              False                True                    False   
2              False                True                    False   
3              False                True                    False   
4              False                True                    False   
...              ...                 ...                      ...   
1673           False                True                    False   
1674           False                True                    False   
1675           False                True                    False   
1676           False                True                    False   
1677           False                True                    False   

      yearswithcurrmanager  sameasmonthlyincome  datebirth  salary  \
0                    False                False      False   False   
1                    False                 True      False   False   
2                    False                 True      False   False   
3                    False                False      False   False   
4                    False                False      False    True   
...                    ...                  ...        ...     ...   
1673                 False                False      False    True   
1674                 False                False      False   False   
1675                 False                 True      False   False   
1676                 False                False      False   False   
1677                 False                 True      False   False   

      roledepartament  numberchildren  remotework  
0                True            True       False  
1                True            True       False  
2               False            True       False  
3                True            True       False  
4                True            True       False  
...               ...             ...         ...  
1673             True            True       False  
1674             True            True       False  
1675             True            True       False  
1676             True            True       False  
1677             True            True       False  

[1678 rows x 10 columns]

In [16]:
# 8. Verifica si hay valores no nulos en el conjunto de datos

# usando info para ver mas datos
df_carmen.info()

# Verificar si hay valores no nulos en todo el DataFrame
df_carmen.notnull()

# Verificar si hay valores no nulos en una columna específica
#df['columna'].notnull()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1678 entries, 0 to 1677
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   yearsatcompany           1678 non-null   int64  
 1   yearsincurrentrole       35 non-null     object 
 2   yearssincelastpromotion  1678 non-null   int64  
 3   yearswithcurrmanager     1678 non-null   int64  
 4   sameasmonthlyincome      1189 non-null   object 
 5   datebirth                1678 non-null   int64  
 6   salary                   1393 non-null   object 
 7   roledepartament          312 non-null    object 
 8   numberchildren           0 non-null      float64
 9   remotework               1678 non-null   object 
dtypes: float64(1), int64(4), object(5)
memory usage: 131.2+ KB


yearsatcompany  yearsincurrentrole  yearssincelastpromotion  \
0               True               False                     True   
1               True               False                     True   
2               True               False                     True   
3               True               False                     True   
4               True               False                     True   
...              ...                 ...                      ...   
1673            True               False                     True   
1674            True               False                     True   
1675            True               False                     True   
1676            True               False                     True   
1677            True               False                     True   

      yearswithcurrmanager  sameasmonthlyincome  datebirth  salary  \
0                     True                 True       True    True   
1                     True                False       True    True   
2                     True                False       True    True   
3                     True                 True       True    True   
4                     True                 True       True   False   
...                    ...                  ...        ...     ...   
1673                  True                 True       True   False   
1674                  True                 True       True    True   
1675                  True                False       True    True   
1676                  True                 True       True    True   
1677                  True                False       True    True   

      roledepartament  numberchildren  remotework  
0               False           False        True  
1               False           False        True  
2                True           False        True  
3               False           False        True  
4               False           False        True  
...               ...             ...         ...  
1673            False           False        True  
1674            False           False        True  
1675            False           False        True  
1676            False           False        True  
1677            False           False        True  

[1678 rows x 10 columns]

In [18]:
# 11. Selecciona las columnas numéricas del conjunto de datos y muestra sus estadísticas descriptivas básicas
columnas_numericas = df_carmen.select_dtypes(include=['number'])
estadisticas_descriptivas = columnas_numericas.describe()
print("\nEstadísticas descriptivas de las columnas numéricas:")
estadisticas_descriptivas


Estadísticas descriptivas de las columnas numéricas:


yearsatcompany  yearssincelastpromotion  yearswithcurrmanager  \
count     1678.000000              1678.000000           1678.000000   
mean         7.124553                 2.241955              4.200238   
std          6.173837                 3.246554              3.554224   
min          0.000000                 0.000000              0.000000   
25%          3.000000                 0.000000              2.000000   
50%          5.000000                 1.000000              3.000000   
75%          9.000000                 3.000000              7.000000   
max         40.000000                15.000000             17.000000   

         datebirth  numberchildren  
count  1678.000000             0.0  
mean   1986.047080             NaN  
std       9.154908             NaN  
min    1963.000000             NaN  
25%    1980.000000             NaN  
50%    1987.000000             NaN  
75%    1993.000000             NaN  
max    2005.000000             NaN

Número de empleados = 1678

Hacer análsis con loc e iloc cuando toque

In [20]:
df_carmen['yearsatcompany']

0       20
1       33
2       22
3       20
4       19
        ..
1673     3
1674     5
1675     6
1676    22
1677     8
Name: yearsatcompany, Length: 1678, dtype: int64

In [22]:
df_carmen['yearsatcompany'].value_counts().sort_index()

yearsatcompany
0      47
1     181
2     145
3     149
4     116
5     214
6      84
7     115
8     113
9      98
10    139
11     37
12     15
13     27
14     19
15     22
16     14
17      9
18     16
19     12
20     30
21     14
22     18
23      2
24      7
25      5
26      4
27      3
29      2
30      1
31      3
32      3
33      5
34      1
36      4
37      1
40      3
Name: count, dtype: int64

In [37]:
#Ver porcentaje por puntuación:
round((df_carmen['yearsatcompany'].value_counts(normalize=True)* 100).sort_index(), 3)

yearsatcompany
0      2.801
1     10.787
2      8.641
3      8.880
4      6.913
5     12.753
6      5.006
7      6.853
8      6.734
9      5.840
10     8.284
11     2.205
12     0.894
13     1.609
14     1.132
15     1.311
16     0.834
17     0.536
18     0.954
19     0.715
20     1.788
21     0.834
22     1.073
23     0.119
24     0.417
25     0.298
26     0.238
27     0.179
29     0.119
30     0.060
31     0.179
32     0.179
33     0.298
34     0.060
36     0.238
37     0.060
40     0.179
Name: proportion, dtype: float64

In [36]:
round(df_carmen['yearsatcompany'].mean(), 3)

7.125

In [38]:
df_datos_abc['numberchildren'].unique()

array([nan])

In [40]:
#15- ¿En que columnas tenemos valores nulos? Utilizad el método que más os guste. 

valores_nulos_por_columna = df_carmen.isna().sum()

# Filtrar para mostrar solo las columnas que tienen valores nulos
columnas_con_nulos = valores_nulos_por_columna[valores_nulos_por_columna > 0]

columnas_con_nulos

yearsincurrentrole     1643
sameasmonthlyincome     489
salary                  285
roledepartament        1366
numberchildren         1678
dtype: int64

In [42]:
#17- ¿Tenemos valores duplicados en el DataFrame?

df_carmen.duplicated().sum()

#si, hay 78 duplicados!

78

In [43]:
df_carmen.columns

Index(['yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')

In [45]:
'''El valor 5 años es el más común (214 personas).
Hay bastantes entre 1 y 10 años.
Luego cae: solo 1 o 2 personas llevan 30, 34 o 37 años.'''

'El valor 5 años es el más común (214 personas).\nHay bastantes entre 1 y 10 años.\nLuego cae: solo 1 o 2 personas llevan 30, 34 o 37 años.'

In [ ]:
# Moda (valor más frecuente):
df_carmen['remotework'].mode()[0]

'1'

In [47]:
# Valores únicos y sus conteos:
df_carmen['roledepartament'].value_counts()  #hay que estandarizar los títulos

roledepartament
rESEArcH SCiEntIst  -  Research & Development            2
MaNAgeR  -  Sales                                        2
hEalthCaRe reprEseNTaTiVe  -  Research & Development     2
SaleS EXeCuTivE  -  Sales                                2
labORAtoRy tEcHNICIAN  -  Research & Development         2
                                                        ..
labORaTORY teChNIcIaN  -  Research & Development         1
saLES eXEcUTiVE  -  Sales                                1
mANUfacTURiNG dIRectOR  -  Research & Development        1
huMAn ResOurces  -  Human Resources                      1
sAleS EXECUtIvE  -  Sales                                1
Name: count, Length: 301, dtype: int64

In [48]:
# Detección de valores no numéricos en columnas que deberían ser numéricas:
# Para la columna 'yearsatcompany'
non_numeric = pd.to_numeric(df_carmen['yearsatcompany'], errors='coerce').isna()
df_carmen[non_numeric]['yearsatcompany'].unique()

array([], dtype=int64)

In [52]:
# Porcentaje de valores nulos:
round((df_carmen.isnull().sum() / len(df_carmen)) * 100, 3)

yearsatcompany               0.000
yearsincurrentrole          97.914
yearssincelastpromotion      0.000
yearswithcurrmanager         0.000
sameasmonthlyincome         29.142
datebirth                    0.000
salary                      16.985
roledepartament             81.406
numberchildren             100.000
remotework                   0.000
dtype: float64

In [54]:
# 5. Conteo de valores únicos por columna
print("\n=== VALORES ÚNICOS ===")
print(df_carmen.nunique())


=== VALORES ÚNICOS ===
yearsatcompany              37
yearsincurrentrole          10
yearssincelastpromotion     16
yearswithcurrmanager        18
sameasmonthlyincome        493
datebirth                   43
salary                     583
roledepartament            301
numberchildren               0
remotework                   5
dtype: int64


In [55]:
# 6. Valores más frecuentes
print("\n=== VALORES MÁS FRECUENTES ===")
for col in df_carmen.columns:
    print(f"\nColumna: {col}")
    print(df_carmen[col].value_counts().head(5))


=== VALORES MÁS FRECUENTES ===

Columna: yearsatcompany
yearsatcompany
5     214
1     181
3     149
2     145
10    139
Name: count, dtype: int64

Columna: yearsincurrentrole
yearsincurrentrole
2,0    12
7,0     5
0,0     4
4,0     3
1,0     3
Name: count, dtype: int64

Columna: yearssincelastpromotion
yearssincelastpromotion
0    655
1    397
2    183
7     95
4     69
Name: count, dtype: int64

Columna: yearswithcurrmanager
yearswithcurrmanager
2    395
0    285
7    275
3    152
8    122
Name: count, dtype: int64

Columna: sameasmonthlyincome
sameasmonthlyincome
2342,59$     228
4492,84$     227
8339,32$     105
12783,92$     42
15943,72$     25
Name: count, dtype: int64

Columna: datebirth
datebirth
1992    89
1988    88
1989    86
1994    82
1987    80
Name: count, dtype: int64

Columna: salary
salary
53914,11$     270
28111,13$     255
100071,84$    122
153407,07$     45
191324,62$     28
Name: count, dtype: int64

Columna: roledepartament
roledepartament
rESEArcH SCiEntIst  - 

In [57]:
# 7. Valores mínimos y máximos (para columnas numéricas)
print("\n=== RANGOS DE VALORES ===")
numeric_cols = df_carmen.select_dtypes(include=['int64', 'float64']).columns
for col in numeric_cols:
    print(f"{col}: Min={df_carmen[col].min()}, Max={df_carmen[col].max()}")


=== RANGOS DE VALORES ===
yearsatcompany: Min=0, Max=40
yearssincelastpromotion: Min=0, Max=15
yearswithcurrmanager: Min=0, Max=17
datebirth: Min=1963, Max=2005
numberchildren: Min=nan, Max=nan


In [58]:
# 8. Detección de valores inusuales (ejemplo para una columna numérica)
print("\n=== VALORES INUSUALES (EJEMPLO) ===")
sample_col = 'yearsatcompany'  # Puedes cambiar por cualquier columna numérica
q1 = df_carmen[sample_col].quantile(0.25)
q3 = df_carmen[sample_col].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - 1.5*iqr
upper_bound = q3 + 1.5*iqr
outliers = df_carmen[(df_carmen[sample_col] < lower_bound) | (df_carmen[sample_col] > upper_bound)]
print(f"Valores inusuales en {sample_col}:")
print(outliers[sample_col].unique())


=== VALORES INUSUALES (EJEMPLO) ===
Valores inusuales en yearsatcompany:
[20 33 22 19 21 24 31 26 23 32 25 37 40 36 27 29 30 34]


Notas importantes:
Algunas columnas parecen tener valores numéricos almacenados como strings (con comas como separadores decimales).

Hay valores booleanos representados de diferentes formas (True/False, Yes/No, 1/0).

Algunas columnas de fechas o años tienen valores no numéricos que deberían limpiarse.



In [61]:
round(df_carmen.isna().sum()/df_carmen.shape[0]*100, 3)

yearsatcompany               0.000
yearsincurrentrole          97.914
yearssincelastpromotion      0.000
yearswithcurrmanager         0.000
sameasmonthlyincome         29.142
datebirth                    0.000
salary                      16.985
roledepartament             81.406
numberchildren             100.000
remotework                   0.000
dtype: float64

In [ ]:
df_carmen.describe()

yearsatcompany  yearssincelastpromotion  yearswithcurrmanager  \
count     1678.000000              1678.000000           1678.000000   
mean         7.124553                 2.241955              4.200238   
std          6.173837                 3.246554              3.554224   
min          0.000000                 0.000000              0.000000   
25%          3.000000                 0.000000              2.000000   
50%          5.000000                 1.000000              3.000000   
75%          9.000000                 3.000000              7.000000   
max         40.000000                15.000000             17.000000   

         datebirth  numberchildren  
count  1678.000000             0.0  
mean   1986.047080             NaN  
std       9.154908             NaN  
min    1963.000000             NaN  
25%    1980.000000             NaN  
50%    1987.000000             NaN  
75%    1993.000000             NaN  
max    2005.000000             NaN

In [62]:
df_carmen.describe().T

count         mean       std     min     25%  \
yearsatcompany           1678.0     7.124553  6.173837     0.0     3.0   
yearssincelastpromotion  1678.0     2.241955  3.246554     0.0     0.0   
yearswithcurrmanager     1678.0     4.200238  3.554224     0.0     2.0   
datebirth                1678.0  1986.047080  9.154908  1963.0  1980.0   
numberchildren              0.0          NaN       NaN     NaN     NaN   

                            50%     75%     max  
yearsatcompany              5.0     9.0    40.0  
yearssincelastpromotion     1.0     3.0    15.0  
yearswithcurrmanager        3.0     7.0    17.0  
datebirth                1987.0  1993.0  2005.0  
numberchildren              NaN     NaN     NaN

### LIMPIEZA

In [67]:
df_carmen.isna().sum()

yearsatcompany                0
yearsincurrentrole         1643
yearssincelastpromotion       0
yearswithcurrmanager          0
sameasmonthlyincome         489
datebirth                     0
salary                      285
roledepartament            1366
numberchildren             1678
remotework                    0
dtype: int64

In [69]:
df_carmen.columns

Index(['yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')

In [73]:
#copia de la variable df_carmen para no borrar nada definitivamente
df_limpio_carmen = df_carmen.copy()
print("Iniciando el proceso de limpieza de datos...")

Iniciando el proceso de limpieza de datos...


In [74]:
#1. Columna 'numberchildren'
# Esta columna no contiene información. Valorar si le recomendamos a la empresa tener o no esa info disponible.
print("Limpiando 'numberchildren'...")
if 'numberchildren' in df_limpio_carmen.columns:
    df_limpio_carmen.drop('numberchildren', axis=1, inplace=True)
    print("'numberchildren' ha sido eliminada.")
else:
    print("'numberchildren' no se encontró en el DataFrame.")

Limpiando 'numberchildren'...
'numberchildren' ha sido eliminada.


In [ ]:
#Comprobamos que se ha eliminado correctamente
df_limpio_carmen.columns

Index(['yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'remotework'],
      dtype='object')

In [76]:
# 2. Columna 'yearsincurrentrole'
# Es de tipo 'object' y los números usan una coma como separador decimal. Tiene una gran cantidad de valores nulos.
print("\nLimpiando 'yearsincurrentrole'...")
df_limpio_carmen['yearsincurrentrole'] = df_limpio_carmen['yearsincurrentrole'].str.replace(',', '.', regex=False).astype(float)
print("'yearsincurrentrole' convertida a tipo numérico (float).")
# NOTA: Esta columna tiene una cantidad muy alta de valores nulos.
# Deberás decidir una estrategia para manejarlos, como rellenarlos (con la media, mediana, etc.) o incluso eliminar la columna si no es útil para tu análisis.



Limpiando 'yearsincurrentrole'...
'yearsincurrentrole' convertida a tipo numérico (float).


In [84]:
df_limpio_carmen.describe().T

count         mean       std     min     25%  \
yearsatcompany           1678.0     7.124553  6.173837     0.0     3.0   
yearsincurrentrole         35.0     3.971429  3.584983     0.0     2.0   
yearssincelastpromotion  1678.0     2.241955  3.246554     0.0     0.0   
yearswithcurrmanager     1678.0     4.200238  3.554224     0.0     2.0   
datebirth                1678.0  1986.047080  9.154908  1963.0  1980.0   

                            50%     75%     max  
yearsatcompany              5.0     9.0    40.0  
yearsincurrentrole          2.0     6.5    13.0  
yearssincelastpromotion     1.0     3.0    15.0  
yearswithcurrmanager        3.0     7.0    17.0  
datebirth                1987.0  1993.0  2005.0

### Ya tenemos yearsatcurrenrole cambiado a float. Falta ver qué hacer con los nulos (amplia mayoría. No le aplicaría la media o mediana puesto que son demasiados nulos y no veo tendencia. La respuesta está en usar yearssincelastpromotion y yearsatcompany.

In [85]:
# 3. Columnas de moneda: 'sameasmonthlyincome' y 'salary'
# Ambas son de tipo 'object', contienen el símbolo '$' y usan comas para los decimales.
print("\nLimpiando 'sameasmonthlyincome' y 'salary'...")

def limpiar_moneda(series_moneda):
    if pd.api.types.is_object_dtype(series_moneda):
        # Aseguramos que todos los valores son strings antes de usar .str
        return series_moneda.astype(str).str.replace('$', '', regex=False).str.replace(',', '.', regex=False).astype(float)
    return series_moneda

df_limpio_carmen['sameasmonthlyincome'] = limpiar_moneda(df_limpio_carmen['sameasmonthlyincome'])
df_limpio_carmen['salary'] = limpiar_moneda(df_limpio_carmen['salary'])
print("Columnas de moneda ('sameasmonthlyincome', 'salary') convertidas a tipo float.")


Limpiando 'sameasmonthlyincome' y 'salary'...
Columnas de moneda ('sameasmonthlyincome', 'salary') convertidas a tipo float.


In [ ]:
df_limpio_carmen.describe().T #ahora también tenemos sameasmonthlyincome y salary como float

count          mean           std      min  \
yearsatcompany           1678.0      7.124553      6.173837      0.0   
yearsincurrentrole         35.0      3.971429      3.584983      0.0   
yearssincelastpromotion  1678.0      2.241955      3.246554      0.0   
yearswithcurrmanager     1678.0      4.200238      3.554224      0.0   
sameasmonthlyincome      1189.0   5676.787065   6157.859086   1000.0   
datebirth                1678.0   1986.047080      9.154908   1963.0   
salary                   1393.0  65070.911587  45849.510802  10090.0   

                              25%       50%       75%       max  
yearsatcompany               3.00      5.00      9.00      40.0  
yearsincurrentrole           2.00      2.00      6.50      13.0  
yearssincelastpromotion      0.00      1.00      3.00      15.0  
yearswithcurrmanager         2.00      3.00      7.00      17.0  
sameasmonthlyincome       2342.59   4492.84   7861.67   94083.0  
datebirth                 1980.00   1987.00   1993.00    2005.0  
salary                   28111.13  53914.11  94390.00  199990.0

In [93]:
df_limpio_carmen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1678 entries, 0 to 1677
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   yearsatcompany           1678 non-null   int64  
 1   yearsincurrentrole       35 non-null     float64
 2   yearssincelastpromotion  1678 non-null   int64  
 3   yearswithcurrmanager     1678 non-null   int64  
 4   sameasmonthlyincome      1189 non-null   float64
 5   datebirth                1678 non-null   int64  
 6   salary                   1393 non-null   float64
 7   roledepartament          312 non-null    object 
 8   remotework               1678 non-null   object 
dtypes: float64(3), int64(4), object(2)
memory usage: 118.1+ KB


In [94]:
# 4. Columna 'roledepartament'
# Tiene espacios en blanco extra y mayúsculas/minúsculas inconsistentes.
print("\nLimpiando 'roledepartament'...")
df_limpio_carmen['roledepartament'] = df_limpio_carmen['roledepartament'].str.strip().str.title()
print("'roledepartament' normalizada (sin espacios extra y en formato Título).")


Limpiando 'roledepartament'...
'roledepartament' normalizada (sin espacios extra y en formato Título).


In [ ]:
df_limpio_carmen['roledepartament'].value_counts() #normalizado, pero hay 1366 filas vacías

roledepartament
Sales Executive  -  Sales                               71
Research Scientist  -  Research & Development           64
Laboratory Technician  -  Research & Development        60
Healthcare Representative  -  Research & Development    27
Manufacturing Director  -  Research & Development       22
Sales Representative  -  Sales                          17
Research Director  -  Research & Development            16
Human Resources  -  Human Resources                     15
Manager  -  Research & Development                      14
Manager  -  Sales                                        5
Manager  -  Human Resources                              1
Name: count, dtype: int64

In [120]:
df_limpio_carmen['roledepartament'].isna().sum()

1366

In [121]:
df_limpio_carmen['roledepartament'].notna().sum()

312

In [137]:
df_3columnas = df_datos_abc[['jobrole', 'department', 'roledepartament']]
df_3columnas

jobrole                department  \
0           resEArch DIREcToR                        NaN   
1                     ManAGeR                        NaN   
2                     ManaGER    Research & Development    
3           ReseArCH DIrECtOr                        NaN   
4             sAleS EXECUtIve                        NaN   
...                        ...                       ...   
1673       rESEaRcH SciEnTiST                        NaN   
1674                  mANager                        NaN   
1675       reSearch sCienTiSt                        NaN   
1676   maNufACTURING DIREctOr                        NaN   
1677       resEArch scIENTisT                        NaN   

                           roledepartament  
0                                      NaN  
1                                      NaN  
2      ManaGER  -  Research & Development   
3                                      NaN  
4                                      NaN  
...                                    ...  
1673                                   NaN  
1674                                   NaN  
1675                                   NaN  
1676                                   NaN  
1677                                   NaN  

[1678 rows x 3 columns]

In [ ]:
df_filtrado = df_3columnas[
    df_3columnas[['jobrole', 'department', 'roledepartament']].notnull().all(axis=1)
]

df_filtrado 

jobrole                department  \
2                        ManaGER    Research & Development    
5                        MANAger    Research & Development    
20     heaLtHcArE repResENtATiVe    Research & Development    
29               saLEs exEcutIVe                     Sales    
30               Sales ExECuTIVe                     Sales    
...                           ...                       ...   
1640             Sales exECutIVE                     Sales    
1646       LaborAtorY Technician    Research & Development    
1654   hEalthCaRe reprEseNTaTiVe    Research & Development    
1658       LABOrATOrY TEchnIcIAn    Research & Development    
1667          rESEArcH SCiEntIst    Research & Development    

                                        roledepartament  
2                   ManaGER  -  Research & Development   
5                   MANAger  -  Research & Development   
20     heaLtHcArE repResENtATiVe  -  Research & Deve...  
29                           saLEs exEcutIVe  -  Sales   
30                           Sales ExECuTIVe  -  Sales   
...                                                 ...  
1640                         Sales exECutIVE  -  Sales   
1646   LaborAtorY Technician  -  Research & Developm...  
1654   hEalthCaRe reprEseNTaTiVe  -  Research & Deve...  
1658   LABOrATOrY TEchnIcIAn  -  Research & Developm...  
1667     rESEArcH SCiEntIst  -  Research & Development   

[312 rows x 3 columns]

### Podemos ver que roledepartament es en realidad una combinación de las columnas jobrole y department, por lo que esta columna es redundante. Sugiero limpiar y completar la columna department (basándonos en jobrole) y borramos roledepartment. La estrategia de nulos aquí es entonces con imputación (dict)

In [145]:
# 5. Columna 'remotework'
# Contiene valores booleanos representados de distintas formas (Yes, 1, True, 0, False).
# Vamos a estandarizarlos a un único tipo booleano (True/False).
print("\nLimpiando 'remotework'...")
mapa_booleano = {
    'Yes': True, '1': True, 1: True, 'True': True, True: True,
    'False': False, '0': False, 0: False, 'No': False, False: False
}
df_limpio_carmen['remotework'] = df_limpio_carmen['remotework'].map(mapa_booleano).astype(bool)
print("'remotework' estandarizada a tipo booleano.")


Limpiando 'remotework'...
'remotework' estandarizada a tipo booleano.


In [146]:
df_limpio_carmen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1678 entries, 0 to 1677
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   yearsatcompany           1678 non-null   int64  
 1   yearsincurrentrole       35 non-null     float64
 2   yearssincelastpromotion  1678 non-null   int64  
 3   yearswithcurrmanager     1678 non-null   int64  
 4   sameasmonthlyincome      1189 non-null   float64
 5   datebirth                1678 non-null   int64  
 6   salary                   1393 non-null   float64
 7   roledepartament          312 non-null    object 
 8   remotework               1678 non-null   bool   
dtypes: bool(1), float64(3), int64(4), object(1)
memory usage: 106.6+ KB


In [152]:
remote_work = df_limpio_carmen["remotework"].value_counts()
remote_work

remotework
True     1042
False     636
Name: count, dtype: int64

In [153]:
# 6. Columnas que ya parecen estar limpias
# 'yearsatcompany', 'yearssincelastpromotion', 'yearswithcurrmanager', 'datebirth'
# Según la información del DataFrame, estas columnas ya son de tipo numérico y no tienen nulos.
# No se requiere una limpieza obvia, pero siempre es bueno revisarlas en busca de valores atípicos.
print("\nLas columnas 'yearsatcompany', 'yearssincelastpromotion', 'yearswithcurrmanager' y 'datebirth' parecen estar ya limpias y no se han modificado.")

print("\n¡Limpieza de datos completada!")


Las columnas 'yearsatcompany', 'yearssincelastpromotion', 'yearswithcurrmanager' y 'datebirth' parecen estar ya limpias y no se han modificado.

¡Limpieza de datos completada!


In [154]:
# --- Verificación Final ---
print("\n--- VERIFICACIÓN DE LA LIMPIEZA ---")
print("\nTipos de datos de las columnas modificadas:")
columnas_modificadas = ['yearsincurrentrole', 'sameasmonthlyincome', 'salary', 'roledepartament', 'remotework']
print(df_limpio_carmen[columnas_modificadas].dtypes)

print("\nPrimeras 5 filas de las columnas limpiadas:")
print(df_limpio_carmen[columnas_modificadas].head())


--- VERIFICACIÓN DE LA LIMPIEZA ---

Tipos de datos de las columnas modificadas:
yearsincurrentrole     float64
sameasmonthlyincome    float64
salary                 float64
roledepartament         object
remotework                bool
dtype: object

Primeras 5 filas de las columnas limpiadas:
   yearsincurrentrole  sameasmonthlyincome    salary  \
0                 NaN             16280.83  195370.0   
1                 NaN                  NaN  199990.0   
2                 NaN                  NaN  192320.0   
3                 NaN             14307.50  171690.0   
4                 NaN             12783.92       NaN   

                      roledepartament  remotework  
0                                 NaN        True  
1                                 NaN        True  
2  Manager  -  Research & Development        True  
3                                 NaN       False  
4                                 NaN       False  


In [155]:
df_limpio_carmen.head()

yearsatcompany  yearsincurrentrole  yearssincelastpromotion  \
0              20                 NaN                       15   
1              33                 NaN                       11   
2              22                 NaN                       11   
3              20                 NaN                        5   
4              19                 NaN                        2   

   yearswithcurrmanager  sameasmonthlyincome  datebirth    salary  \
0                    15             16280.83       1972  195370.0   
1                     9                  NaN       1971  199990.0   
2                    15                  NaN       1981  192320.0   
3                     6             14307.50       1976  171690.0   
4                     8             12783.92       1977       NaN   

                      roledepartament  remotework  
0                                 NaN        True  
1                                 NaN        True  
2  Manager  -  Research & Development        True  
3                                 NaN       False  
4                                 NaN       False

In [158]:
df_limpio_carmen.isna().sum()

yearsatcompany                0
yearsincurrentrole         1643
yearssincelastpromotion       0
yearswithcurrmanager          0
sameasmonthlyincome         489
datebirth                     0
salary                      285
roledepartament            1366
remotework                    0
dtype: int64

In [165]:
df_income = df_datos_abc[["monthlyincome", "sameasmonthlyincome"]]
df_income

monthlyincome sameasmonthlyincome
0        16280,83$           16280,83$
1              NaN                 NaN
2              NaN                 NaN
3        14307,50$           14307,50$
4        12783,92$           12783,92$
...            ...                 ...
1673      3949,17$            3949,17$
1674     15943,72$           15943,72$
1675           NaN                 NaN
1676      8339,32$            8339,32$
1677           NaN                 NaN

[1678 rows x 2 columns]

In [169]:

df_salary = df_datos_abc[["salary", "joblevel"]]
df_salary

salary  joblevel
0     195370,00$         5
1     199990,00$         5
2     192320,00$         5
3     171690,00$         4
4            NaN         4
...          ...       ...
1673         NaN         1
1674  191324,62$         5
1675   28111,13$         1
1676  100071,84$         3
1677   53914,11$         2

[1678 rows x 2 columns]

In [174]:
df_salary.describe().T

count      mean       std  min  25%  50%  75%  max
joblevel  1678.0  2.064362  1.099425  1.0  1.0  2.0  3.0  5.0

In [ ]:
df_salary.describe(include="O").T #me sale como object porque no está limpia la columna dentro de df_datos_abc (en carmen sí)

count unique        top freq
salary  1393    583  53914,11$  270